### PR TITLE
[ios] Synchronize the app light/dark appearance with the device

### PR DESCRIPTION
**Pull request by kirylkaveryn**
See https://github.com/organicmaps/organicmaps/issues/749 , not closing it yet as Android part should also be implemented.

### In this PR:
1. Synchronize the auto-switching between light/dark themes to the device when the user selects `Auto` in the OM Settings.
When the On/Off is selected in OM Settings synchronization is disabled.
2. When the user uses CarPlay and Carplay auto-updates content style then the map theme will be changed to it. There is no possibility to force the map theme whiole using the CarPlay (except Carplay's settings). 
3. A lot of bugs related to the auto theme changing were fixed, because with the current `ThemeManager` and `StyleManager` systems UI elements usually support the creation with the current theme, but do not auto-update when the theme is changing.
4. Move the NightMode selection from the `Navigation` section into `General`

Tested on: 
iOS 15.5, 17.2,
iPhone 15 pro, iPoneSE 3gen, iPad, iPad Designed For Mac

Simulator Screen Recording - iPhone SE (3rd generation) - 2024-02-01 at 14 07 04

![Simulator Screen Recording - iPhone 15 Pro - 2024-01-31 at 18 43 07]
Simulator Screen Recording - iPhone 15 Pro - 2024-02-01 at 13 31 42
Simulator Screen Recording - iPhone SE (3rd generation) - 2024-02-01 at 13 28 33
(https://github.com/organicmaps/organicmaps/assets/79797627/8f8d0aba-f131-4fd4-86e1-0bcd1dfb0b46)

https://github.com/organicmaps/organicmaps/assets/79797627/b9187811-d6b8-49af-bc35-1fc9f7c7e65f

CarPlay:
Auto
Main-VideoStream-2024-01-31-14-42-42

DarkMapsDisabled
Main-VideoStream-2024-01-31-14-41-52
Main-VideoStream-2024-01-31-14-43-05

DarkMapsEnabled
Main-VideoStream-2024-01-31-14-42-55
Main-VideoStream-2024-01-31-14-43-00

### CODE REVIEW

File: iphone/Maps/Classes/CarPlay/CarPlayService.swift [hunk 1]
```diff
@@ -40,19 +40,6 @@ final class CarPlayService: NSObject {
...
-    if #available(iOS 13.0, *) {
-      switch configuration.contentStyle {
-      case .light:
-        rootTemplateStyle = .light
-        window.overrideUserInterfaceStyle = .light
-      case .dark:
-        rootTemplateStyle = .dark
-        window.overrideUserInterfaceStyle = .dark
-      default:
-        rootTemplateStyle = window.overrideUserInterfaceStyle == .light ? .light : .dark
```
Comment (biodranik) [inline iphone/Maps/Classes/CarPlay/CarPlayService.swift:53]: Was this case ever used? Is it handled now?

File: iphone/Maps/Classes/CustomAlert/BaseAlert/MWMAlert.mm [hunk 2]
```diff
@@ -266,4 +266,21 @@ - (void)layoutSubviews {
...
+  [super traitCollectionDidChange:previousTraitCollection];
+  if (self.traitCollection.userInterfaceStyle != previousTraitCollection.userInterfaceStyle) {
+    [self updateViewStyle:self];
+  }
+}
+
+- (void)updateViewStyle:(UIView *)view {
+  if (!view)
+    return;
+  view.isStyleApplied = NO;
```
Comment (biodranik) [inline iphone/Maps/Classes/CustomAlert/BaseAlert/MWMAlert.mm:279]: How is this variable used?

File: iphone/Maps/Core/Settings/MWMSettings.mm [hunk 3]
```diff
@@ -92,6 +92,23 @@ + (void)setTheme:(MWMTheme)theme
...
     return;
+  if (@available(iOS 13.0, *)) {
+    UIUserInterfaceStyle style;
+    switch (theme) {
+      case MWMThemeDay:
+      case MWMThemeVehicleDay:
+        style = UIUserInterfaceStyleLight;
+        break;
+      case MWMThemeNight:
+      case MWMThemeVehicleNight:
```
Comment (biodranik) [inline iphone/Maps/Core/Settings/MWMSettings.mm:103]: Is Outdoor theme related to these values? How is this enum used, and can it be simplified? Is it needed at all?

File: iphone/Maps/Core/Theme/Core/ThemeManager.swift [hunk 4]
```diff
@@ -28,7 +18,8 @@ final class ThemeManager: NSObject {
       case .night: fallthrough
       case .vehicleNight: return isVehicleRouting ? .vehicleNight : .night
       case .auto:
-        if #available(iOS 13.0, *), ThemeManager.enableDarkModeBySystem {
+        let isDarkModeEnabled = UIApplication.shared.delegate?.window??.traitCollection.userInterfaceStyle == .dark
```
Comment (biodranik) [inline iphone/Maps/Core/Theme/Core/ThemeManager.swift:21]: Can this line be moved under the if below?

File: iphone/Maps/UI/Settings/MWMSettingsViewController.mm [hunk 5]
```diff
@@ -196,7 +180,23 @@ - (void)configNavigationSection {
       break;
   }
   [self.nightModeCell configWithTitle:L(@"pref_map_style_title") info:nightMode];
+}
+
+- (void)show3dBuildingsAlert:(UITapGestureRecognizer *)recognizer {
+  UIAlertController *alert =
```
Comment (biodranik) [inline iphone/Maps/UI/Settings/MWMSettingsViewController.mm:186]:
```suggestion
  UIAlertController * alert =
```

File: iphone/Maps/UI/Settings/MWMSettingsViewController.mm [hunk 6]
```diff
@@ -196,7 +180,23 @@ - (void)configNavigationSection {
...
+  [UIAlertController alertControllerWithTitle:L(@"pref_map_3d_buildings_title")
```
Comment (biodranik) [inline iphone/Maps/UI/Settings/MWMSettingsViewController.mm:187]: Usually indent of 4 spaces is used in these cases. Without it, the line looks like a new, separate operator.

File: iphone/Maps/UI/Settings/MWMSettingsViewController.mm [hunk 7]
```diff
@@ -196,7 +180,23 @@ - (void)configNavigationSection {
...
+                                      message:L(@"pref_map_3d_buildings_disabled_summary")
+                               preferredStyle:UIAlertControllerStyleAlert];
+
+  UIAlertAction *okButton = [UIAlertAction actionWithTitle:@"OK"
```
Comment (biodranik) [inline iphone/Maps/UI/Settings/MWMSettingsViewController.mm:191]:
```suggestion
  UIAlertAction * okButton = [UIAlertAction actionWithTitle:@"OK"
```

File: iphone/Maps/Classes/CarPlay/CarPlayService.swift (hunk 1 above)
Comment (kirylkaveryn) [reply iphone/Maps/Classes/CarPlay/CarPlayService.swift:53]: We should not use `window.userInterfaceStyle` - this property is responsible for the car play UI,  not `content`. Important to separate this terms. 
We should only response to the contentTheme (templates with arrows, estimates etc). 
Content theme have only 2 cases: dark and light.

image

File: iphone/Maps/Classes/CustomAlert/BaseAlert/MWMAlert.mm (hunk 2 above)
Comment (kirylkaveryn) [reply iphone/Maps/Classes/CustomAlert/BaseAlert/MWMAlert.mm:279]: Uneuseful chunk of code! (copy-pasted from StyleManager).  
thx!

File: iphone/Maps/Core/Settings/MWMSettings.mm (hunk 3 above)
Comment (kirylkaveryn) [reply iphone/Maps/Core/Settings/MWMSettings.mm:103]: 1. Outdoor is not part of this enum. It is used only in Core managers.
image

2. it is hard to say now is these cases MWMThemeVehicleNight/day are still useful... needs additional refactoring. 
Core enum contains them:
image

Should I refactor these cases?

File: iphone/Maps/UI/Settings/MWMSettingsViewController.mm (hunk 6 above)
Comment (kirylkaveryn) [reply iphone/Maps/UI/Settings/MWMSettingsViewController.mm:187]: In this file, I don't change anything except moving this chunk from one method into another (NightMode from one section to another). 
Should I fix the formatting on the file In this commit?

Comment (biodranik) [reply iphone/Maps/UI/Settings/MWMSettingsViewController.mm:187]: It's ok to fix some minor issues if you're already editing that file. And avoid copy-pasting the entropy )

File: iphone/Maps/Core/Settings/MWMSettings.mm (hunk 3 above)
Comment (biodranik) [reply iphone/Maps/Core/Settings/MWMSettings.mm:103]: _Kostyli_ detected, let's refactor them later.

File: iphone/Maps/Classes/CarPlay/CarPlayService.swift (hunk 1 above)
Comment (biodranik) [reply iphone/Maps/Classes/CarPlay/CarPlayService.swift:53]: It used to work fine in CarPlay before. Hope this change won't break it.

**Review by biodranik — APPROVED** [review]
LGTM after minor comments are fixed.

File: iphone/Maps/Classes/CarPlay/CarPlayService.swift (hunk 1 above)
Comment (kirylkaveryn) [reply iphone/Maps/Classes/CarPlay/CarPlayService.swift:53]: I followed the documentation and tested a lot. Think it should work well!

> The vehicle selects the content style according to the ambient light level. Your navigation app can use this value to determine the most appropriate style of **map content to draw** in its base view. The **content style is independent of the user interface style**, which controls light and dark mode.

https://developer.apple.com/documentation/carplay/cpsessionconfiguration/contentstyle?language=objc

**Review by biodranik — APPROVED** [review]
Thanks! Let's see it in action!

**Pull request merged**